### PR TITLE
design: Improved Stash Migrations

### PR DESCRIPTION
**Pull request by parker-timmerman**
Design doc to improve Stash migrations. Any input or comments is much appreciated!

### Motivation

#17466 

### Tips for reviewer

[Rendered](https://github.com/MaterializeInc/materialize/pull/18881/files?short_path=cf7216a#diff-cf7216ab01653f49e7e5045403d65fa7952f4dd2ca238a66c103aea35421556d)

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

File: doc/developer/design/19042023_stash_migrations.md
Comment (jkosh44) [inline doc/developer/design/19042023_stash_migrations.md:89]: When does this value get updated in the stash?

File: doc/developer/design/19042023_stash_migrations.md
Comment (jkosh44) [inline doc/developer/design/19042023_stash_migrations.md:248]: What about the existing upgrade tests that we have? Would those work?

https://github.com/MaterializeInc/materialize/tree/main/test/upgrade

File: doc/developer/design/19042023_stash_migrations.md
Comment (jkosh44) [inline doc/developer/design/19042023_stash_migrations.md:77]: Where does `migrate.rs` fit into this proposal?

https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/migrate.rs

File: doc/developer/design/19042023_stash_migrations.md
Comment (jkosh44) [inline doc/developer/design/19042023_stash_migrations.md:208]: I think this implies that the only type of migrations we support are from versions `X` to `X + 1`, and not arbitrary migrations from `X` to `X + n`. That sounds correct to me, and consistent with our current approach. I just think we should explicitly say that somewhere in this doc.

Comment (jkosh44) [reply doc/developer/design/19042023_stash_migrations.md:208]: Additionally if that's the case, then we may want an assert somewhere here that `version` is equal to `STASH_VERSION` or `STASH_VERSION - 1`.

File: doc/developer/design/19042023_stash_migrations.md
Comment (jkosh44) [inline doc/developer/design/19042023_stash_migrations.md:99]: Currently, the first migration is used to set up the stash with initial values.

https://github.com/MaterializeInc/materialize/blob/b38572937e354e93de330a2a36ee76ab676edf3e/src/adapter/src/catalog/storage.rs#L86-L424

Where does that initialization step fit into this proposal?

File: doc/developer/design/19042023_stash_migrations.md
Comment (benesch) [inline doc/developer/design/19042023_stash_migrations.md:338]: It does not! I feel very strongly that nothing but the one `environmentd` in charge should be looking at a given stash, with an exception for emergency debugging tools (`stash-debug`). If you want to know what's in a stash, ask `environmentd`; don't look in the stash directly. We changed the design of billing to adhere to this principle.

File: doc/developer/design/19042023_stash_migrations.md
Comment (benesch) [inline doc/developer/design/19042023_stash_migrations.md:352]: FWIW Cockroach uses Snappy to compress data before it hits the disk, so additional compression likely won't move the needle on how much disk storage is required. It could in theory reduce the amount of data we're transiting to Cockroach over pgwire but .. I think a massive `CREATE VIEW` statement is out of spec for Materialize. I'd be willing to bet that folks are using massive `CREATE VIEW`s with static data only because our table support is limited, and big `VALUES` clauses are a way to work around that.

File: doc/developer/design/19042023_stash_migrations.md
Comment (benesch) [inline doc/developer/design/19042023_stash_migrations.md:232]: Are we able to remove old migrations after they have shipped in a release? IME it is very important for the long term maintainability of a system to be able to retire old migrations, rather than letting them accrete forever.

**Review by benesch — COMMENTED** [review]
Only had time for a quick skim, but excited to see us tackle this! Argument for using protobuf over the alternatives checks out for me.

**Review by mjibson — COMMENTED** [review]
Overall I think this looks good. Some possible concerns (not well thought through yet).

1. How easy will it be to tell what changed at each migration? git diff adjacent .proto files? Snapshotting has some benefits, but easily observing changes might not be one.
2. The above becomes important when we're trying to verify that a migration was written correctly. When reading PRs, will we be able to observe the diff of the current migration because that file is actually being changed? That seems ok if so.
3. Could we ever squash these, or should we just...not ever do it? In some far future point it'd be nice if a new customer didn't have to wait more than a second or so for all of the migrations to run. Probably not an issue now.
4. Can you add a section about the initial migration to this? If the migrations are all written assuming they started at an empty v0, we gotta figure out what to do for our current users.

File: doc/developer/design/19042023_stash_migrations.md
Comment (parker-timmerman) [reply doc/developer/design/19042023_stash_migrations.md:93]: I'm definitely being a bit "handwavey" at the moment with transactions and what not, but we would update the version after a successful migration from N to N + 1. I updated the doc to add a specific "set version" step

File: doc/developer/design/19042023_stash_migrations.md
Comment (parker-timmerman) [reply doc/developer/design/19042023_stash_migrations.md]: Yes, those would actually be great! I updated the doc to include these

File: doc/developer/design/19042023_stash_migrations.md
Comment (parker-timmerman) [reply doc/developer/design/19042023_stash_migrations.md:81]: AFAICT `migrate.rs` more so handles migrations of the Catalog, in other words, we've read in all of the data from the Stash and now treat everything as SQL objects, and operate on them as so. Whereas Stash migrations operate on "primitive" data, e.g. Strings, ints, and bools, and have no context to how they might relate to SQL objects.

IMO both code paths can and should exist separately, because they operate on the data at different layers.

File: doc/developer/design/19042023_stash_migrations.md
Comment (parker-timmerman) [reply doc/developer/design/19042023_stash_migrations.md:229]: Good callout, let me add this!

File: doc/developer/design/19042023_stash_migrations.md
Comment (parker-timmerman) [reply doc/developer/design/19042023_stash_migrations.md]: Added this to the design doc as it's own section, tl;dr I propose we add a new "initialize" step to the Stash which populates the Stash with the necessary data, and sets the Stash version to the `STASH_VERSION` we have in code.

File: doc/developer/design/19042023_stash_migrations.md
Comment (parker-timmerman) [reply doc/developer/design/19042023_stash_migrations.md]: Great! I totally agree with the idea that `environmentd` should be the only application code that interacts with the Stash, let me call this out in the doc

File: doc/developer/design/19042023_stash_migrations.md
Comment (parker-timmerman) [reply doc/developer/design/19042023_stash_migrations.md:410]: That totally makes sense, I'll call this out in the doc, thanks!

File: doc/developer/design/19042023_stash_migrations.md
Comment (parker-timmerman) [reply doc/developer/design/19042023_stash_migrations.md:253]: This totally makes sense, and I believe we should be able to remove old migrations. I added a new section on an "initialization" step for the Stash, which should allow us to remove old migrations, and maintain a "minimum stash version" that we support migrating from.

**Comment by parker-timmerman** [comment]
> 1. How easy will it be to tell what changed at each migration? git diff adjacent .proto files? Snapshotting has some benefits, but easily observing changes might not be one.

I believe it should be fairly easy. The flow for snapshotting would be:
1. Make the necessary changes to `objects.proto`. This should show up as a normal diff, and thus it should be easy to see what changed.
2. Copy the existing `objects.proto` file to `objects_vX.proto`. GitHub should identify this as a copy and mark the file as such, e.g. it will say something like "copied from `objects.proto`"

> The above becomes important when we're trying to verify that a migration was written correctly. When reading PRs, will we be able to observe the diff of the current migration because that file is actually being changed? That seems ok if so.

Yes you should be able to

>  Could we ever squash these, or should we just...not ever do it? In some far future point it'd be nice if a new customer didn't have to wait more than a second or so for all of the migrations to run. Probably not an issue now.

Yes I added a new section on this! Specifically moving what is today's v0 migration that we use to initialize the stash, to its own "initialization step". That should allow us to deprecate versions of the stash so new customers don't need to wait.


>  Can you add a section about the initial migration to this? If the migrations are all written assuming they started at an empty v0, we gotta figure out what to do for our current users.

**Review by jkosh44 — APPROVED** [review]

**Comment by benesch** [comment]
(Don't wait for my approval! Not planning to review in depth. I'm happy as long as there is a plan for vacuuming up old migrations.)

**Comment by mjibson** [comment]
> Specifically moving what is today's v0 migration that we use to initialize the stash, to its own "initialization step". That should allow us to deprecate versions of the stash so new customers don't need to wait.

The new section on the initial migration looks good. However I don't see any words about allowing us to deprecate old versions of the stash. Is that implied somehow?

File: doc/developer/design/19042023_stash_migrations.md
Comment (ParkMyCar) [inline doc/developer/design/19042023_stash_migrations.md:282]: @mjibson here is a small blurb on deprecating old migrations. The idea is that once we've released the migration, in the next release of the binary we could bump `MIN_STASH_VERSION` and remove the old migration path, similar to what we do today.

Comment (mjibson) [reply doc/developer/design/19042023_stash_migrations.md:282]: Ah ok! LGTM then, should be ok to merge.

**Pull request merged**
